src/client/ds3/user.c

```diff
@@ -0,0 +1,185 @@
+/**
+ * (C) Copyright 2022 Intel Corporation.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause-Patent
+ */
+
+#include "ds3_internal.h"
+
+int
+ds3_user_set(const char *name, struct ds3_user_info *info, ds3_t *ds3, daos_event_t *ev)
+{
+	int         rc;
+	dfs_obj_t  *user_obj;
+	mode_t      mode = DEFFILEMODE;
+	d_sg_list_t wsgl;
+	d_iov_t     iov;
+	char       *user_path;
+	int         i;
+
+	if (ds3 == NULL || name == NULL || info == NULL)
+		return -EINVAL;
+
+	// Remove old user data
+	ds3_user_remove(name, info, ds3, ev);
+
+	// Open user file
+	rc = dfs_open(ds3->meta_dfs, ds3->meta_dirs[USERS_DIR], name, S_IFREG | mode,
+		      O_RDWR | O_CREAT | O_TRUNC, 0, 0, NULL, &user_obj);
+	if (rc != 0) {
+		D_ERROR("Failed to open user file, name = %s, rc = %d\n", name, rc);
+		goto err_ret;
+	}
+
+	// Write user data
+	d_iov_set(&iov, info->encoded, info->encoded_length);
+	wsgl.sg_nr   = 1;
+	wsgl.sg_iovs = &iov;
+	rc           = dfs_write(ds3->meta_dfs, user_obj, &wsgl, 0, ev);
+	dfs_release(user_obj);
+	if (rc != 0) {
+		D_ERROR("Failed to write to user file, name = %s, rc = %d\n", name, rc);
+		goto err_ret;
+	}
+
+	// Build user path
+	D_ALLOC_ARRAY(user_path, DS3_MAX_KEY);
+	strcpy(user_path, "../");
+	strcat(user_path, meta_dir_name(USERS_DIR));
+	strcat(user_path, "/");
+	strcat(user_path, name);
```

Comment:
Is there validation elsewhere that these can't add up to more than `DS3_MAX_KEY`?